Couldn't import the type definitions of std_msgs in my Typescript program

**Description**

I seem to be unable to import the type definitions of std_msgs in my Typescript program. I am new to typescript/javascript, but I think I've checked the obvious stuff:

My `tsconfig.json` file contains the following:
```
...
  "compilerOptions": {
   ...
    "module": "commonjs",
    "moduleResolution": "node",
    "target": "es6"
    },
...
```

I ran `npx generate-ros-messages` and in `<yourproject>/node_modules/rclnodejs/generated/` I see a `std_msgs` directory with files for every type, including strings.
I also have an `interfaces.d.ts` file located in `<yourproject>/node_modules/types`, and it contains:
```
...
declare module 'rclnodejs' {
...
  namespace std_msgs {
  ...
    namespace msgs {
    ...
      export interface String {
      ...
```

I'm just not sure why I can't access the `rclnodejs.std_msgs.msg.String` type directly, do I have to separately include interface.d.ts?

- Library Version:   "rclnodejs": "0.27.4"
- ROS Version: Ros2 - Humble
- Platform / OS: Description:    Ubuntu 20.04.6 LTS

**Steps To Reproduce**
1. Start the following typescript code using `npx tsx src/example.ts`
2. Send a ros2 message to the /msg_sub topic using `ros2 topic pub /msg_sub std_msgs/msg/String "data: hello"`


<!--- Include the minimum possible code or example data to reproduce the problem -->

```typescript
import * as rclnodejs from 'rclnodejs';

async function main() {
  await rclnodejs.init();

  const node = new rclnodejs.Node('ros2_pub_ex');
  const publisher = node.createPublisher('std_msgs/msg/String', 'msg_pub');
  node.createSubscription('std_msgs/msg/String', 'msg_sub', (msg) => {
    if ( msg instanceof rclnodejs.std_msgs.msg.String) {
      console.log(`Received std_msgs/msg/String with data: ${msg.data}`)
    }
  });
  node.spin();

  let num = 0;
  setInterval(() => {
      let stringMsgObject = rclnodejs.createMessageObject('std_msgs/msg/String');
      stringMsgObject.data = `hello world ${num++}`;
      console.log('publishing ', stringMsgObject);
      publisher.publish(stringMsgObject);
  }, 1000);
}

void main();

process.on('SIGINT', () => process.exit(1));
``` 


**Expected Behavior**

I expected a console printout of receiving the message. Ex:
`Received std_msgs/msg/String with data: hello`

**Actual Behavior**

```shell
/workspaces/rclnodejs_example/src/example.ts:10
    if ( msg instanceof rclnodejs.std_msgs.msg.String) {
                                           ^

TypeError: Cannot read properties of undefined (reading 'msg')
```

Thanks for the response! I updated that section of the code:

```typescript
  node.createSubscription('std_msgs/msg/String', 'msg_sub', (msg) => {
    console.log('msg recieved ',msg);
    if ( msg instanceof rclnodejs.require('std_msgs').msg.String) {
      console.log(`Received std_msgs/msg/String with data: ${msg.data}`)
    }
  });
```

And when I run the typescript compiler `npx tsc --project ./tsconfig.json` I get the following build error:

```shell

error TS2339: Property 'msg' does not exist on type 'object'.

10     if ( msg instanceof rclnodejs.require('std_msgs').msg.String) {
                                                         ~~~

error TS2339: Property 'data' does not exist on type 'Buffer<ArrayBufferLike> | String'.
  Property 'data' does not exist on type 'Buffer<ArrayBufferLike>'.

11       console.log(`Received std_msgs/msg/String with data: ${msg.data}`)
                                                                    ~~~~
```

The code still compiles to js, and I can run it with `node ./dist/example.js`, but if I run it and then in a separate terminal, run `ros2 topic pub /msg_sub std_msgs/msg/String "data: hello"`, I get the printout that the message is received
```
msg recieved  { data: 'hello' }
```
but not the printout about the data field
```
Received std_msgs/msg/String with data: ${msg.data}
```

I could work around the error by casting the type of `msg` as `any`, but that defeats the purpose of using typescript
```typescript
  node.createSubscription('std_msgs/msg/String', 'msg_sub', (msg:any) => {
    console.log('msg received ',msg);
    console.log(`Received std_msgs/msg/String with data: ${msg.data}`)
  });
```

I did some more work on this, and I can't tell the callback to expect the type `rclnodejs.std_msgs.msg.String` directly, resulting in the same error as before
```typescript
  node.createSubscription('std_msgs/msg/String', 'msg_sub', (msg:rclnodejs.std_msgs.msg.String) => {
    console.log('msg received ',msg);
    console.log(`Received std_msgs/msg/String with data: ${msg.data}`)
  });
```
```shell
error TS2345: Argument of type '(msg: rclnodejs.std_msgs.msg.String) => void' is not assignable to parameter of type 'SubscriptionCallback<"std_msgs/msg/String">'.
  Types of parameters 'msg' and 'message' are incompatible.
    Type 'Buffer<ArrayBufferLike> | String' is not assignable to type 'String'.
      Property 'data' is missing in type 'Buffer<ArrayBufferLike>' but required in type 'String'.
```

But, I was able to figure out that I could get the build error to go away if I first expect `msg` to be of type `rclnodejs.Message`, and then assert it to be of type `rclnodejs.std_msgs.msg.String`
```typescript
  node.createSubscription('std_msgs/msg/String', 'msg_sub', (msg:rclnodejs.Message) => {
    console.log('msg received ',msg);
    console.log(`Received std_msgs/msg/String with data: ${(msg as rclnodejs.std_msgs.msg.String).data}`)
  });
```

Ah, my mistake, I see what happened here. I'm using version 0.27.4, and issue https://github.com/RobotWebTools/rclnodejs/issues/1013 led to a fix in 0.29.0. However, when I try upgrading to version 0.29.0 (or any version >= 0.28.1), I get the following error when I try to run `npm install rclnodejs@0.29.0`

```shell
npm error g++: error: unrecognized command line option ‘-std=c++20’; did you mean ‘-std=c++2a’?
npm error make: *** [rclnodejs.target.mk:163: Release/obj.target/rclnodejs/src/addon.o] Error 1
```

This indicates that rclnodejs v0.28.1 or greater requires gcc10. I'm running this in a container running Ubuntu 20.04, which defaults to gcc version 9.4.0. So, either I try to install gcc10 into the container, or I upgrade the container to Ubuntu 22.04. I'll update once I've done that.

--- a/binding.gyp
+++ b/binding.gyp
@@ -12,6 +12,7 @@
     'runtime%': 'node',
     'ros_lib_dir': "<!(node -p \"require('./scripts/config.js').getROSLibPath()\")",
     'ros_include_root': "<!(node -p \"require('./scripts/config.js').getROSIncludeRootPath()\")",
+    'node_major_version': '<!(node -p \"process.versions.node.split(\'.\')[0]\")',
   },
   'targets': [
     {
@@ -62,8 +63,21 @@
             'defines': [
               'OS_LINUX'
             ],
-            'cflags_cc': [
-              '-std=c++20'
+            'conditions': [
+              [
+                'node_major_version >= 23', {
+                  'cflags': [
+                    '-std=c++20'
+                  ]
+                }
+              ],
+              [
+                'node_major_version < 23', {
+                  'cflags_cc': [
+                    '-std=c++17'
+                  ]
+                }
+              ]
             ]
           }
         ],
@@ -72,8 +86,21 @@
             'defines': [
               'OS_WINDOWS'
             ],
-            'cflags_cc': [
-              '-std=c++20'
+            'conditions': [
+              [
+                'node_major_version >= 23', {
+                  'cflags': [
+                    '-std=c++20'
+                  ]
+                }
+              ],
+              [
+                'node_major_version < 23', {
+                  'cflags_cc': [
+                    '-std=c++17'
+                  ]
+                }
+              ]
             ],
             'include_dirs': [
               './src/third_party/dlfcn-win32/',

--- a/binding.gyp
+++ b/binding.gyp
@@ -66,7 +66,7 @@
             'conditions': [
               [
                 'node_major_version >= 23', {
-                  'cflags': [
+                  'cflags_cc': [
                     '-std=c++20'
                   ]
                 }
@@ -89,7 +89,7 @@
             'conditions': [
               [
                 'node_major_version >= 23', {
-                  'cflags': [
+                  'cflags_cc': [
                     '-std=c++20'
                   ]
                 }

